uber/models.py

```diff
@@ -1791,6 +1799,60 @@ def html(self):
             return SafeString(self.body.replace('\n', '<br/>'))
 
 
+class PromoCode(MagModel):
+    expiration_date = Column(UTCDateTime, default=c.ESCHATON)
+    price = Column(Integer, default=0)
+    code = Column(UnicodeText)
+    old_code = ""
+    # Uses should be -1 to make expired
+    uses = Column(Integer, nullable=True, default=None)
+    used_by = relationship(Attendee)
+
+    def generate_code(self, count):
+        code = []
+        for x in range(count):
+            code.append(c.PROMO_CODE_WORDS[random.choice(string.ascii_letters + string.digits)])
+        return "_".join(code)
+
+    @presave_adjustment
+    def _usage_count(self):
+        if self.uses == 0:
+            self.uses = None
+
+    @presave_adjustment
+    def _check_code(self):
+        if self.old_code == "":
+            if self.code == '':
+                self.code = self.generate_code(6)
+        elif self.old_code != self.code:
+            with Session() as session:
+                while True:
+                    match = session.query(PromoCode).filter(PromoCode.code == self.code and PromoCode.id != self.id).first()
+                    if match:
+                        split_code = self.code.split("_")
+                        split_code.append(self.generate_code(1))
+                        self.code = "_".join(split_code)
+                    else:
+                        self.old_code = self.code
+                        break
+
+    def use(self, user):
+        if not self.expired:
+            self.used_by.append(user)
+
+    def expire(self):
+        self.expiration_date = datetime.now(UTC)
```

Comment:
Almost everything in the rest of the codebase uses a custom function called `localized_now`. Even though it's not super necessary to have this be in local time, it's probably better for future coders if the expiration date is always in localized time.